### PR TITLE
Feature/vivox #19

### PR DESCRIPTION
**Pull request by zhengHugo**
Adding vivox core scripts.

To compile, you need to import vivox asset to your local environment:
1. Search for Vivox in Unity asset store, the one with a size of more than 300MB. 
2. Before clicking the import button, make sure to EXCLUDE the folder "Mirror" and "Sample", and "Test" is optional. 
3. Click import

### CODE REVIEW

**Comment by zhengHugo** [comment]
For testing: 
1. Build VivoxScene under Yuguo's scene sandbox and run.
2. Login with an arbitrary username. 
3. Click "join channel"
4. Run another instance of the same scene, login and join channel with another username
5. You should be able to hear your own voice now.

**Comment by eli-lam** [comment]
^ I suggest you to not include the Sample and Test folders in your import, they contain errors

**Comment by eli-lam** [comment]
@zhengHugo I'm getting this error. Any way to fix? 
image

**Comment by zhengHugo** [comment]
> @zhengHugo I'm getting this error. Any way to fix?
> image

Can you try import Vivox again? It looks like you didn't import x86_64 package in the "Plugins" folder. 
Also, make sure to exclude "Mirror", "Sample", and "Test"

File: .gitignore [hunk 1]
```diff
@@ -61,6 +61,5 @@ sysinfo.txt
 # Crashlytics generated file
 crashlytics-build.properties
 
-# Vivox
+# Vivox 
 /Assets/Vivox/
```
Comment (eli-lam) [inline .gitignore:65]: Can you change this to `/Assets/Vivox*`

Comment (zhengHugo) [reply .gitignore]: Done

**Review by eli-lam — APPROVED** [review]
Can you prefix all private members with `this.` when they are being called?

**Comment by zhengHugo** [comment]
> > @zhengHugo I'm getting this error. Any way to fix?
> > image
> 
> Can you try to import Vivox again? It looks like you didn't import x86_64 package in the "Plugins" folder.
> Also, make sure to exclude "Mirror", "Sample", and "Test"

Solution to this:
1. In build settings, add the VivoxScene
2. Select the build platform based on your OS

**Comment by zhengHugo** [comment]
> Can you prefix all private members with `this.` when they are being called?

Fixed.
Thank you for your review!

**Review by shifatkhan — CHANGES_REQUESTED** [review]
I can hear my own voice! 😱
Works perfectly fine, good stuff!

Can you just change the functions with 1 line of code into lambdas?
e.g. `private int Square(int x) => x*x;`

File: Assets/Scripts/Vivox/UI/TextChatUI.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    private void Awake()
+    {
+        this.vivoxManager = VivoxManager.Instance;
+        this.vivoxManager.OnTextMessageLogReceivedEvent += OnTextMessageLogReceivedEvent;
+    }
+
+    private void OnDestroy()
+    {
+        this.vivoxManager.OnTextMessageLogReceivedEvent -= OnTextMessageLogReceivedEvent;
+    }
```
Comment (shifatkhan) [inline Assets/Scripts/Vivox/UI/TextChatUI.cs:29]: lamda:
```csharp
private void OnDestroy() => this.vivoxManager.OnTextMessageLogReceivedEvent -= OnTextMessageLogReceivedEvent;
```

File: Assets/Scripts/Vivox/VivoxManager.cs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        {
+            this.client.Uninitialize();
+            this.client.Initialize();
+            DontDestroyOnLoad(this);
+        }
+
+        private void OnApplicationQuit()
+        {
+            this.client.Uninitialize();
+        }
```
Comment (shifatkhan) [inline Assets/Scripts/Vivox/VivoxManager.cs:80]: lamda:
```csharp
private void OnApplicationQuit() => this.client.Uninitialize();
```

File: Assets/Scripts/Vivox/VivoxManager.cs [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            else
+            {
+                Debug.LogError("[Vivox] Cannot join a channel when not logged in.");
+            }
+        }
+
+        public void LeaveChannel(IChannelSession channelSession)
+        {
+            channelSession.Disconnect();
+        }
```
Comment (shifatkhan) [inline Assets/Scripts/Vivox/VivoxManager.cs:187]: lamda:
```csharp
 public void LeaveChannel(IChannelSession channelSession) => channelSession.Disconnect();
 ```

File: Assets/Scripts/Vivox/VivoxManager.cs [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        // bind with UI Login Button
+        public void BtnLogin()
+        {
+            LogIn();
+        }
+
+        public void BtnJoinChannel()
+        {
+            JoinChannel(true, true, ChannelType.NonPositional);
+        }
```
Comment (shifatkhan) [inline Assets/Scripts/Vivox/VivoxManager.cs:257]: lamda:
```csharp
public void BtnLogin() => LogIn();
```

and 
```csharp
public void BtnJoinChannel() => JoinChannel(true, true, ChannelType.NonPositional);
```

File: Assets/Scripts/Vivox/VivoxManager.cs [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    }
+}
```
Comment (shifatkhan) [inline Assets/Scripts/Vivox/VivoxManager.cs:259]: Add a line break

File: Assets/Scripts/Vivox/UI/TextChatUI.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        if (string.IsNullOrEmpty(this.messageInputField.text))
+        {
+            return;
+        }
+
+        this.vivoxManager.SendTextMessage(this.messageInputField.text);
+        ClearOutTextField();
+    }
+}
```
Comment (shifatkhan) [inline Assets/Scripts/Vivox/UI/TextChatUI.cs:55]: Add a line break

**Review by shifatkhan — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Vivox/UI/TextChatUI.cs (hunk 7 above)
Comment (zhengHugo) [reply Assets/Scripts/Vivox/UI/TextChatUI.cs]: Fixed. Thanks for your review.

File: Assets/Scripts/Vivox/UI/TextChatUI.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+/// Text Chat UI in game scene
+/// </summary>
+public class TextChatUI : MonoBehaviour
+{
+    private VivoxManager vivoxManager;
+
+    [SerializeField] private Button sendButton;
+    [SerializeField] private InputField messageInputField;
+    [SerializeField] private Text newMessageText;
+
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Vivox/UI/TextChatUI.cs:18]: don't need this extra space

File: Assets/Scripts/Vivox/UI/TextChatUI.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        IChannelTextMessage channelTextMessage)
+    {
+        var text = channelTextMessage.Message;
+        this.newMessageText.text = text;
+    }
+
+    public void BtnSendText()
+    {
+        if (string.IsNullOrEmpty(this.messageInputField.text))
+        {
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Vivox/UI/TextChatUI.cs:47]: can lose the brackets for this one liner

**Review by danielwiktorczyk — APPROVED** [review]
Looks good! I know many of these files are boilerplate from Vivox themselves, so I won't make you refactor their code to suit our conventions, haha

**Review by shifatkhan — APPROVED** [review]
Good stuff !

**Pull request merged**